build/rpmfc.c

```diff
@@ -1180,22 +1180,42 @@ static int initAttrs(rpmfc fc)
 {
     ARGV_t files = NULL;
     char * attrPath = rpmExpand("%{_fileattrsdir}/*.attr", NULL);
-    int nattrs = 0;
+    int nfiles = 0;
+    int nlocals = 0;
+    int attrcnt = 0;
+    int j;
 
     /* Discover known attributes from pathnames + initialize them */
     if (rpmGlob(attrPath, NULL, &files) == 0) {
-	nattrs = argvCount(files);
-	fc->atypes = xcalloc(nattrs + 1, sizeof(*fc->atypes));
-	for (int i = 0; i < nattrs; i++) {
-	    char *bn = basename(files[i]);
-	    bn[strlen(bn)-strlen(".attr")] = '\0';
-	    fc->atypes[i] = rpmfcAttrNew(bn);
+	nfiles = argvCount(files);
+    }
+    char * local_attr_names = rpmExpand("%{?__local_file_attrs}", NULL);
+    ARGV_t local_attrs = argvSplitString(local_attr_names, ":", ARGV_SKIPEMPTY);
+    nlocals = argvCount(local_attrs);
+    fc->atypes = xcalloc(nfiles + nlocals + 1, sizeof(*fc->atypes));
+
+    for (int i = 0; i < nfiles; i++) {
+	char *bn = basename(files[i]);
+	bn[strlen(bn)-strlen(".attr")] = '\0';
+	// skip file attrs found in __local_file_attrs for now
+	for (j=0; (j < nlocals) && strcmp(bn, local_attrs[j]); j++);
+	if (j < nlocals) {
+	    free(bn);
```

Comment:
This is not right, basename() doesn't allocate, so this would free some random pointer inside the one that is.

Reply:
Might be more readable to just toss all basenames into a new argv with argvAddUniq() , which frees us from having to care about these details here.